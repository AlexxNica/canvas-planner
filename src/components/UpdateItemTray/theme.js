/* Global variables (colors, typography, spacing, etc.) are defined in lib/themes */

export default function generator ({ colors, typography }) {
  const cssVars = {
    background: colors.white
  };
  return cssVars;
}
