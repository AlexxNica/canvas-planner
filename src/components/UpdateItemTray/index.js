import React, { Component } from 'react';
import _ from 'lodash';
import themeable from 'instructure-ui/lib/themeable';
import Container from 'instructure-ui/lib/components/Container';
import Button from 'instructure-ui/lib/components/Button';
import formatMessage from '../../format-message';
import PropTypes from 'prop-types';
import TextInput from 'instructure-ui/lib/components/TextInput';
import Select from 'instructure-ui/lib/components/Select';
import TextArea from 'instructure-ui/lib/components/TextArea';
import IconCalendarMonthSolid from 'instructure-icons/react/Solid/IconCalendarMonthSolid';
import Grid, {GridRow, GridCol} from 'instructure-ui/lib/components/Grid';

import styles from './styles.css';
import theme from './theme.js';

export class UpdateItemTray extends Component {
  static propTypes = {
    courses: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.string,
      longName: PropTypes.string,
    })).isRequired,
    noteItem: PropTypes.object,
    onSavePlannerItem: PropTypes.func.isRequired,
    onDeletePlannerItem: PropTypes.func.isRequired,
  };

  constructor (props) {
    super(props);
    const updates = _.cloneDeep(props.noteItem) || {};
    if (updates.context) {
      updates.courseId = updates.context.id;
      delete updates.context;
    }
    this.state = {
      updates,
      titleMessages: [],
      dateMessages: [],
    };
  }

  handleSave = () => {
    const updates = Object.assign({}, this.state.updates);
    if (updates.courseId) {
      updates.context = { id: updates.courseId };
    }
    delete updates.courseId;
    this.props.onSavePlannerItem(updates);
  }

  handleChange = (field, value) => {
    this.setState({
      updates: {
        ...this.state.updates,
        [field]: value
      }
    });
  }

  handleCourseIdChange = (e) => {
    let value = e.target.value;
    if (value === 'none') value = undefined;
    this.handleChange('courseId', value);
  }

  handleTitleChange = (e) => {
    const value = e.target.value;
    if (value === '') {
      this.setState({
        titleMessages: [{type: 'error', text: formatMessage('title is required')}]
      });
    } else {
        this.setState({titleMessages: []});
    }
    this.handleChange('title', value);
  }

  handleDateChange = (e) => {
    const value = e.target.value;
    if (value === '') {
      this.setState({dateMessages: [{type: 'error', text: formatMessage('date is required')}]});
    } else {
      this.setState({dateMessages: []});
    }
    this.handleChange('date', value);
  }

  handleDeleteClick = () => {
    this.props.onDeletePlannerItem(this.props.noteItem);
  }

  findCurrentValue (field) {
    return this.state.updates[field] || '';
  }

  isValid () {
    // intentional empty string checking
    return this.state.updates.title && this.state.updates.date;
  }

  renderDeleteButton () {
    if (this.props.noteItem == null) return;
    return <Button
      variant="light"
      margin="0 x-small 0 0"
      onClick={this.handleDeleteClick}>
      {formatMessage("Delete")}
    </Button>;
  }

  renderSaveButton () {
    return <Button
      variant="primary"
      disabled={!this.isValid()}
      onClick={this.handleSave}>
        {formatMessage("Save")}
    </Button>;
  }

  renderTitleInput () {
    const value = this.findCurrentValue('title');
    return (
      <TextInput
        label={formatMessage("Title")}
        value={value}
        messages={this.state.titleMessages}
        onChange={this.handleTitleChange}
      />
    );
  }

  renderDateInput () {
    const value = this.findCurrentValue('date');
    return (
      <TextInput
        label={formatMessage("Date")}
        value={value}
        messages={this.state.dateMessages}
        onChange={this.handleDateChange}
      />
    );
  }

  renderCourseSelectOptions () {
    if (!this.props.courses) return [];
    return this.props.courses.map(course => {
      return <option key={course.id} value={course.id}>{course.longName}</option>;
    });
  }

  renderCourseSelect () {
    let courseId = this.findCurrentValue('courseId');
    if (courseId == null) courseId = 'none';
    return (
      <Select
        label={formatMessage("Course")}
        value={courseId}
        onChange={this.handleCourseIdChange}
      >
        <option value="none">{formatMessage("Optional: Add Course")}</option>
        {this.renderCourseSelectOptions()}
      </Select>
    );
  }

  renderDetailsInput () {
    const value = this.findCurrentValue('details');
    return (
      <TextArea
        label={formatMessage("Details")}
        height="12rem"
        autoGrow={false}
        value={value}
        onChange={(e) => this.handleChange('details', e.target.value)}
      />
    );
  }

  render () {
    return (
      <div className={styles.root}>
        <Container
          display="block"
          as="section"
          margin="large medium"
          textAlign="end"
        >
          <Container
            display="block"
            margin="medium 0"
          >
            {this.renderTitleInput()}
          </Container>
          <Grid vAlign="bottom" colSpacing="none">
            <GridRow>
              <GridCol>
                {this.renderDateInput()}
              </GridCol>
              <GridCol>
                <Button variant="icon">
                  <IconCalendarMonthSolid />
                </Button>
              </GridCol>
            </GridRow>
          </Grid>
          <Container
            display="block"
            margin="medium 0"
          >
            {this.renderCourseSelect()}
          </Container>
          <Container
            display="block"
            margin="medium 0"
          >
            {this.renderDetailsInput()}
          </Container>
          {this.renderDeleteButton()}
          {this.renderSaveButton()}
        </Container>
      </div>
    );
  }
}

export default themeable(theme, styles)(UpdateItemTray);
