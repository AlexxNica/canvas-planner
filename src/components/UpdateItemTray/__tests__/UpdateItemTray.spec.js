import React from 'react';
import { shallow, mount } from 'enzyme';
import 'instructure-ui/lib/themes/canvas';
import UpdateItemTray from '../index';

it('renders the base component correctly with one buttons and four inputs', () => {
  const wrapper = shallow(
    <UpdateItemTray />
  );
  expect(wrapper).toMatchSnapshot();
});

it('renders the item to update if provided', () => {
  const noteItem = {
    title: 'Planner Item',
    date: '2017-04-25 01:49:00-0700',
    courseId: '1',
    details: "You made this item to remind you of something, but you forgot what."
  };
  const wrapper = shallow(
    <UpdateItemTray noteItem={noteItem} />
  );
  expect(wrapper).toMatchSnapshot();
});

it('shows title inputs', () => {
  const wrapper = mount(
    <UpdateItemTray />
  );
  expect(wrapper.find('TextInput')).toHaveLength(2);
  const input = wrapper.find('TextInput').first();
  input.find('input').simulate('change', {target: {value: 'New Text'}});
  expect(input.props().value).toEqual('New Text');
});

it('shows details inputs', () => {
  const wrapper = mount(
    <UpdateItemTray />
  );
  const input = wrapper.find('TextArea');
  input.find('textarea').simulate('change', {target: {value: 'New Details'}});
  expect(input.props().value).toEqual('New Details');
});
