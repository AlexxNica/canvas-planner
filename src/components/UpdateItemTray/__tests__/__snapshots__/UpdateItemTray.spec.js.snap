// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the item to update if provided 1`] = `
<div>
  <Container
    as="section"
    display="block"
    margin="large medium"
    textAlign="end"
  >
    <Container
      display="block"
      margin="medium 0"
    >
      <TextInput
        disabled={false}
        inline={false}
        inputRef={[Function]}
        label="Title"
        layout="stacked"
        messages={Array []}
        onChange={[Function]}
        size="medium"
        type="text"
        value="Planner Item"
      />
    </Container>
    <Grid
      colSpacing="none"
      hAlign="start"
      rowSpacing="medium"
      startAt="small"
      vAlign="bottom"
      visualDebug={false}
    >
      <GridRow>
        <GridCol
          textAlign="inherit"
        >
          <DateInput
            defaultDateValue="2017-04-25 01:49:00-0700"
            format="LL"
            label="Date"
            locale="en"
            nextLabel="Next Month"
            onDateChange={[Function]}
            previousLabel="Previous Month"
            timeZone="America/Denver"
            validationFeedback={true}
          />
        </GridCol>
      </GridRow>
    </Grid>
    <Container
      display="block"
      margin="medium 0"
    >
      <Select
        disabled={false}
        inline={false}
        label="Course"
        layout="stacked"
        messages={Array []}
        onChange={[Function]}
        selectRef={[Function]}
        size="medium"
        type="text"
        value=""
      >
        <option
          value="none"
        >
          Optional: Add Course
        </option>
        <option
          value="1"
        >
          a course
        </option>
      </Select>
    </Container>
    <Container
      display="block"
      margin="medium 0"
    >
      <TextArea
        autoGrow={false}
        disabled={false}
        height="12rem"
        inline={false}
        label="Details"
        layout="stacked"
        messages={Array []}
        onChange={[Function]}
        resize="none"
        size="medium"
        textareaRef={[Function]}
        value="You made this item to remind you of something, but you forgot what."
      />
    </Container>
    <Button
      as="button"
      buttonRef={[Function]}
      fluidWidth={false}
      margin="0 x-small 0 0"
      onClick={[Function]}
      size="medium"
      type="button"
      variant="light"
    >
      Delete
    </Button>
    <Button
      as="button"
      buttonRef={[Function]}
      disabled={false}
      fluidWidth={false}
      margin="0"
      onClick={[Function]}
      size="medium"
      type="button"
      variant="primary"
    >
      Save
    </Button>
  </Container>
</div>
`;
